src/csharp/Grpc.Core/Channel.cs

```diff
@@ -69,20 +69,28 @@ public Channel(string target, ChannelCredentials credentials, IEnumerable<Channe
             EnsureUserAgentChannelOption(this.options);
             this.environment = GrpcEnvironment.AddRef();
 
-            this.completionQueue = this.environment.PickCompletionQueue();
-            using (var nativeChannelArgs = ChannelOptions.CreateChannelArgs(this.options.Values))
+            try
             {
-                var nativeCredentials = credentials.ToNativeCredentials();
-                if (nativeCredentials != null)
-                {
-                    this.handle = ChannelSafeHandle.CreateSecure(nativeCredentials, target, nativeChannelArgs);
-                }
-                else
+                this.completionQueue = this.environment.PickCompletionQueue();
+                using (var nativeChannelArgs = ChannelOptions.CreateChannelArgs(this.options.Values))
                 {
-                    this.handle = ChannelSafeHandle.CreateInsecure(target, nativeChannelArgs);
+                    var nativeCredentials = credentials.ToNativeCredentials();
+                    if (nativeCredentials != null)
+                    {
+                        this.handle = ChannelSafeHandle.CreateSecure(nativeCredentials, target, nativeChannelArgs);
+                    }
+                    else
+                    {
+                        this.handle = ChannelSafeHandle.CreateInsecure(target, nativeChannelArgs);
+                    }
                 }
+                GrpcEnvironment.RegisterChannel(this);
+            }
+            catch (Exception e)
+            {
+                Task.Run(async () => await GrpcEnvironment.ReleaseAsync().ConfigureAwait(false));
```

Comment:
`this.environment` from `GrpcEnvironment.AddRef()` isn't passed to `GrpcEnvironment.ReleaseAsync()`? Is it just reference counting?

Reply:
yes, basically. GrpcEnvironment is a singleton with global state of grpc native (and e.g. owns the threadpool).